### PR TITLE
Ceph(fix): Updates given to StorageClass won't work.

### PR DESCRIPTION
**Pull request by rajatsingh25aug**
This PR fixes the issue, where the updates given on StorageClass were not working. Now, the cluster will delete the existing sc and create a new one when requested for updation.

Signed-off-by: RAJAT SINGH <rajasing@redhat.com>

This PR is dependent on #563 and will rebase it when that will get merged.
BZ id: https://bugzilla.redhat.com/show_bug.cgi?id=1846085

### CODE REVIEW

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 1]
```diff
@@ -65,22 +65,22 @@ func (r *ReconcileStorageCluster) createStorageClasses(scs []*storagev1.StorageC
...
-			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+		if err == nil || errors.IsNotFound(err) {
+			if err == nil {
+				if existing.DeletionTimestamp != nil {
+					reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
+					return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				}
+				reqLogger.Info("Deleteing StorageClass %s", existing.Name)
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/initialization_reconciler.go:74]:
```suggestion
				reqLogger.Info("Deleting StorageClass %s", existing.Name)
```

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 2]
```diff
@@ -65,22 +65,22 @@ func (r *ReconcileStorageCluster) createStorageClasses(scs []*storagev1.StorageC
...
-		case err == nil:
-			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+		if err == nil || errors.IsNotFound(err) {
+			if err == nil {
+				if existing.DeletionTimestamp != nil {
+					reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
+					return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				}
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/initialization_reconciler.go:73]: This error message is misleading. We are checking for DeletionTimestamp on StorageClass not Initialization Object.

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 3]
```diff
@@ -65,22 +65,22 @@ func (r *ReconcileStorageCluster) createStorageClasses(scs []*storagev1.StorageC
 		existing := storagev1.StorageClass{}
 		err := r.client.Get(context.TODO(), types.NamespacedName{Name: sc.Name, Namespace: sc.Namespace}, &existing)
 
-		switch {
-		case err == nil:
-			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+		if err == nil || errors.IsNotFound(err) {
+			if err == nil {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/initialization_reconciler.go:69]: Why not use `switch..case` or `if..else if..` instead of nesting?

**Comment by rajatsingh25aug** [comment]
@umangachapagain , Thanks for the review, how can I check **iff** there's an update for the existing SC?. Thanks

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 1 above)
Comment (rajatsingh25aug) [reply pkg/controller/storagecluster/initialization_reconciler.go]: Done!

**Comment by rajatsingh25aug** [comment]
I will resolve the PR once #563 is merged.

**Comment by humblec** [comment]
This need a backport PR to `release-4.5` branch. @rajatsingh25aug @umangachapagain @jarrpa

**Comment by umangachapagain** [comment]
`ci/prow/verify-latest-csv` is failing and `ci/prow/ocs-operator-e2e-aws` was re-triggered. how did this get merged?

**Comment by rajatsingh25aug** [comment]
@umangachapagain `ci/prow/verify-latest-csv` fails sometimes and needs to be re-triggered in order to run it successfully.
Here's a screenshot of `verify-latest-csv` when triggered on my machine.
image

**Comment by umangachapagain** [comment]
> @umangachapagain `ci/prow/verify-latest-csv` fails sometimes and needs to be re-triggered in order to run it successfully.
>
I'd expect that to be done.

> Here's a screenshot of `verify-latest-csv` when triggered on my machine.
>
If it passed on local machine, it should pass on CI too.

My point being, we shouldn't merge without understanding why the CI failed. And, if we are sure it's a constantly reappearing issue in CI, we need to fix it.

**Comment by rajatsingh25aug** [comment]
> > @umangachapagain `ci/prow/verify-latest-csv` fails sometimes and needs to be re-triggered in order to run it successfully.
> 
> I'd expect that to be done.
> 
> > Here's a screenshot of `verify-latest-csv` when triggered on my machine.
> 
> If it passed on local machine, it should pass on CI too.
> 
> My point being, we shouldn't merge without understanding why the CI failed. And, if we are sure it's a constantly reappearing issue in CI, we need to fix it.

Sure, Understood. Also, `e2e` CI (`ci/prow/red-hat-storage-ocs-ci-e2e-aws` and `ci/prow/ocs-operator-e2e-aws` )
fails all the time. And I've seen that on almost all the PRs.

**Pull request merged**